x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/io/stream/PlanNamedTypes.java

```diff
@@ -310,6 +311,13 @@ public static List<PlanNameRegistry.Entry> namedTypeEntries() {
             of(EsField.class, EsField.class, PlanNamedTypes::writeEsField, PlanNamedTypes::readEsField),
             of(EsField.class, DateEsField.class, PlanNamedTypes::writeDateEsField, PlanNamedTypes::readDateEsField),
             of(EsField.class, InvalidMappedField.class, PlanNamedTypes::writeInvalidMappedField, PlanNamedTypes::readInvalidMappedField),
+            of(EsField.class, MultiTypeEsField.class, PlanNamedTypes::writeMultiTypeEsField, PlanNamedTypes::readMultiTypeEsField),
+            of(
+                EsField.class,
+                MultiTypeEsField.UnresolvedField.class, // This extends InvalidMappedField with information needed only during Analysis
+                PlanNamedTypes::writeInvalidMappedField,
+                PlanNamedTypes::readInvalidMappedField
+            ),
```

Comment:
Agreed. It did actually fix a bunch of things, because `UnresolveUnionTypes` was becoming quite complex in order to get regressions to pass, and putting this here made it much simpler. However, if I instead combine the classes `InvalidMappedField` and `MultiTypeEsField.UnresolvedField` into one class, that same simplification occurs, without this hack. I noticed a lot of failures in CI that I could not reproduce locally, but this particular hack feels like a likely source of the failures.

Reply:
Also, in case it was not clear, the existence of this in the plan was only when there is no AbstractConvertFunction, so the remaining InvalidMappedField needs to be serialised in the plan, to support existing behaviour around unsupported types.